### ISSUE TITLE
PlayerExperienceChangeEvent

### ISSUE THREAD

**Issue by Heisenburger69** [pmmp/PocketMine-MP#3182]
## Description
$event->getOldXp() and $event->getNewXp();

## Justification
The above methods would be pretty useful to calculate the change in experience of a player

**Comment by dktapps** [comment]
https://github.com/pmmp/PocketMine-MP/blob/stable/src/pocketmine/entity/Human.php#L418

**Comment by Heisenburger69** [comment]
and how would I get the XP the player had before the Event? I want to get the change in XP and getCurrentTotalXp() would only give me the Player's experience after the event

**Comment by jasonwynn10** [comment]
> and how would I get the XP the player had before the Event? I want to get the change in XP and getCurrentTotalXp() would only give me the Player's experience after the event

False. The function dylan showed above gives you the total xp before the event because the event applies after the listeners are called.

**Comment by Heisenburger69** [comment]
then what about the XP after the Event haha?

**Comment by jasonwynn10** [comment]
did you even look at any of the available methods?? https://github.com/pmmp/PocketMine-MP/blob/stable/src/pocketmine/event/player/PlayerExperienceChangeEvent.php

**Comment by Heisenburger69** [comment]
getNewLevel() returns the Player's level, getNewProgress() returns the progress within the level

**Comment by Heisenburger69** [comment]
getNewLevel() returns the Player's level, getNewProgress() returns the progress within the level

**Comment by jasonwynn10** [comment]
And that is all the event handles, so why should there be an arbitrary function in the event to calculate unused information?

**Comment by Heisenburger69** [comment]
because calculating the raw change in experience itself rather than just level/progress might be a useful addition? For Example: If the change in experience itself is a multiple of 10, double the experience the player gains

**Comment by jasonwynn10** [comment]
The API already exists elsewhere, so it would just be boilerplate code.
https://github.com/pmmp/PocketMine-MP/blob/stable/src/pocketmine/entity/utils/ExperienceUtils.php

**Comment by Heisenburger69** [comment]
getXpToReachLevel doesn't account for in-level progress and hence wouldn't give an accurate value of the change in XP

**Comment by SOF3** [comment]
Instead of having the conversion trouble, why not simply have an `Experience` class that internally stores level+progress and exposes getter method for total XP, in addition to arithmetic methods?

**Comment by Heisenburger69** [comment]
or just add the methods I suggested since that'd be easier to use, especially since change implies well, change in raw experience as well so it'd make sense for the event to have those methods

**Comment by dktapps** [comment]
You can use the code in Human->getCurrentTotalXp() and adapt it for post-event calculation.

The reason this doesn't already exist is because it's possible for arbitrary changes to occur to XP and level which are not locked to a points scale. For example, enchanting does not deduct an amount of XP, it deducts an amount of levels, irrespective of the work needed to complete those levels. In addition, these changes can result in the player having a fractional amount of XP points.

**Comment by JackMD** [comment]
Here's how I do it. Hope this helps

```php
public function onExpGain(PlayerExperienceChangeEvent $event){
	/** @var Player $player */
	$player = $event->getEntity();

	$oldExp = $player->getCurrentTotalXp();

	$newLevel = $event->getNewLevel();
	$newProgress = $event->getNewProgress();

	$remainderExp = (int) (ExperienceUtils::getXpToCompleteLevel($newLevel) * $newProgress);
	$newExp = ExperienceUtils::getXpToReachLevel($newLevel) + $remainderExp;

	$diff = $newExp - $oldExp;

	var_dump("old: $oldExp", "new: $newExp", "diff: $diff");
}
```
